### PR TITLE
registry/storagedriver S3 Walk optimization

### PR DESCRIPTION
**Pull request by CollinShoop**
### **Objective**
blobstore enumeration with S3 storage driver (and possibly others with follow up effort) can be optimized by several orders of magnitude in most cases by offloading more work to the S3 API. 

### **Changes**
 - Use `ListObjectsV2PagesWithContext` **without** `Delimiter`, giving all files in a depth-first recursive ordering in batches up to 1000
 - Infer directories from object paths. keep track of the previous directory and infer all directories in-between that need to be walked over
 - Keep track of the most recent skipped directory and manually skip over any results under that directory
   - **Note**: This does have a potential to be locally less efficient for non-trivial directories. The overall performance increase should more than make up for this, though. 
 - Added tests for `WalkFallback`
 - Added tests for S3 Walk implementation

### **S3 Walk Test**
```
export REGION_ENDPOINT=sfo3.digitaloceanspaces.com
export AWS_REGION=us-east-1
export AWS_ACCESS_KEY=<key>
export AWS_SECRET_KEY=<secret>
export S3_BUCKET=<bucket>
go test -run TestWalk -v ./registry/storage/driver/s3-aws/...
```

### **Comparison** (TODO need to redo)
For each registry ran `BlobEnumerator::enumerate` twice: Once before making these chances & again with the changes. Local changes were made to track the progress in the S3 `Walk` and `WalkFiles` implementations, giving the took, API calls, objects, and folders values below. 

**Test 1 (medium)**
- `WalkFiles` 
  - Took 0.128 seconds
  - 1 API call
  - 298 objects 
  - 0 folders
- `Walk` 
   - Took 53.9 seconds
   - 479 API calls
   - 298 objects 
   - 478 folders

**Results**
- Reduce runtime by factor of 421
- Reduce API calls by a factor of 479 

**Test 2 (large)**
Only the first 5 minutes of `Walk` are recorded and extrapolated, which I think is fair to get the point across
- `WalkFiles` 
  - Took 9.85 seconds
  - 49 API call
  - 48101 objects 
  - 0 folders
- `Walk` partial results
   - Took 5 minutes
   - 2680 API calls
   - 2663 objects (~1/18)
   - 2827 folders
- `Walk` extrapolated (x18)
   - Took 89.74 minutes
   - 48100 API calls
   - 47795 objects
   - 50739 folders

**Results**
- Reduce runtime by factor of 546
- Reduce API calls by a factor of 981.

### CODE REVIEW

File: registry/storage/driver/s3-aws/s3.go [hunk 1]
```diff
@@ -1164,28 +1116,44 @@ func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, pre
...
 					retError = err
 					return false
 				}
 			}
+
+			err := f(walkInfo)
+			*objectCount++
+
+			if err != nil {
+				if err == storagedriver.ErrSkipDir {
```
Comment (waynr) [inline registry/storage/driver/s3-aws/s3.go:1151]: Is it possible that `ErrSkipDir` is returned for an actual object? In that case, what was the previous behavior? Seems to me like we would want to do the same thing as when this err is returned for a "directory".

Comment (CollinShoop) [reply registry/storage/driver/s3-aws/s3.go:1152]: Following what is documented, 
```
	// Walk traverses a filesystem defined within driver, starting
	// from the given path, calling f on each file.
	// If the returned error from the WalkFn is ErrSkipDir and fileInfo refers
	// to a directory, the directory will not be entered and Walk
	// will continue the traversal.  If fileInfo refers to a normal file, processing stops
	Walk(ctx context.Context, path string, f WalkFn) error
```
and based on the impl of `WalkFallback` I'm interpreting `ErrSkipDir` on a file to mean "break" aka graceful stop. So actually, should be returning `false` to stop the S3 list.

File: registry/storage/driver/walk.go [hunk 2]
```diff
@@ -22,9 +22,14 @@ type WalkFn func(fileInfo FileInfo) error
 // to a directory, the directory will not be entered and Walk
 // will continue the traversal.  If fileInfo refers to a normal file, processing stops
 func WalkFallback(ctx context.Context, driver StorageDriver, from string, f WalkFn) error {
+	err, _ := doWalkFallback(ctx, driver, from, f)
+	return err
+}
+
+func doWalkFallback(ctx context.Context, driver StorageDriver, from string, f WalkFn) (error, bool) {
```
Comment (CollinShoop) [inline registry/storage/driver/walk.go:29]: Turns out the fallback implementation actually wasn't properly handling `If fileInfo refers to a normal file, processing stops` as found with a unit test. Added a bool return that can be used to gracefully stop the Walk.

File: registry/storage/driver/s3-aws/s3.go [hunk 3]
```diff
@@ -1164,28 +1116,46 @@ func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, pre
 			})
 		}
 
-		sort.SliceStable(walkInfos, func(i, j int) bool { return walkInfos[i].FileInfoFields.Path < walkInfos[j].FileInfoFields.Path })
-
 		for _, walkInfo := range walkInfos {
```
Comment (waynr) [inline registry/storage/driver/s3-aws/s3.go:1119]: I would add a comment before this loop explaining that directory recursion is unnecessary under the assumption that `ListObjectsV2` returns objects in deterministically-sorted order with an explanation of how we make use of that property to skip directories where possible. It would probably be a good idea to quote from and link to this to justify that assumption: https://docs.aws.amazon.com/AmazonS3/latest/userguide/ListingKeysUsingAPIs.html

**Review by waynr — APPROVED** [review]
Looks good to me, other than the nit about explaining the principle of the new `doWalk` implementation.

File: registry/storage/driver/s3-aws/s3.go (hunk 3 above)
Comment (CollinShoop) [reply registry/storage/driver/s3-aws/s3.go:1129]: Added 👍🏻

**Comment by CollinShoop** [comment]
Finding through further testing, the current impl does not work for scan repositories so addressing that now.

File: registry/storage/driver/s3-aws/s3.go [hunk 4]
```diff
@@ -1116,12 +1117,28 @@ func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, pre
...
+			dirs := directoryDiff(prevDir, filePath)
+			if len(dirs) > 0 {
+				for _, dir := range dirs {
+					walkInfos = append(walkInfos, storagedriver.FileInfoInternal{
+						FileInfoFields: storagedriver.FileInfoFields{
+							IsDir: true,
+							Path:  dir,
+						},
+					})
+					prevDir = dir
```
Comment (waynr) [inline registry/storage/driver/s3-aws/s3.go:1132]: If there is more than one element in `dirs`, would we end up with duplicate `walkInfos` elements for some of those directories since only the last one gets assigned to `prevDir` and passed in to `directoryDiff` on the next object? So, supposing we had two deeply nested files:

```
/meow/hello/world.txt
/meow/hello/world/foo/bar/baz/bonk.txt
/meow/hello/world/foo/bar/baz/boink.txt
```

If I understand correctly with this would result in the following `walkInfos`:

* `/meow/hello`
* `/meow/hello/world/`
* `/meow/hello/world/foo`
* `/meow/hello/world/foo/bar`
* `/meow/hello/world/foo/bar/baz`
* `/meow/hello/world/foo/bar/baz/bonk`
* `/meow/hello/world/foo`
* `/meow/hello/world/foo/bar`
* `/meow/hello/world/foo/bar/baz`
* `/meow/hello/world/foo/bar/baz/boink`

Comment (CollinShoop) [reply registry/storage/driver/s3-aws/s3.go:1132]: `prevDir = dir` is done intentionally as a simple way to save the last dir, which ends up being the parent of the object (or is not re-assigned if directoryDiff yields empty results). This doesn't cause any duplicates though, the unit tests should cover a good amount of cases. 

```
directoryDiff("/meow/hello", "/meow/hello/world/foo/bar/baz/bonk.txt") 
   => ["/meow/hello/world", "/meow/hello/world/foo", "/meow/hello/world/foo/bar", "/meow/hello/world/foo/bar/baz"]
```
results in `prevDir = "/meow/hello/world/foo/bar/baz"`

Then next iteration 
```
directoryDiff("/meow/hello/world/foo/bar/baz", "/meow/hello/world/foo/bar/baz/boink.txt") 
   => [ ]
``` 
because they both belong to the same directory, no new directory `walkInfo` are added. `prevDir` is left as ` "/meow/hello/world/foo/bar/baz"`

File: registry/storage/driver/s3-aws/s3.go [hunk 5]
```diff
@@ -1181,6 +1179,38 @@ func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, pre
...
+//   => [ "/path/to/folder/folder" ],
+// Eg 2 directoryDiff("/path/to/folder/folder1", "/path/to/folder/folder2/file")
+//   => [ "/path/to/folder/folder2" ]
+// Eg 3 directoryDiff("/path/to/folder/folder1/file", "/path/to/folder/folder2/file")
+//  => [ "/path/to/folder/folder2" ]
+// Eg 4 directoryDiff("/path/to/folder/folder1/file", "/path/to/folder/folder2/folder1/file")
+//   => [ "/path/to/folder/folder2", "/path/to/folder/folder2/folder1" ]
+// Eg 5 directoryDiff("/", "/path/to/folder/folder/file")
+//   => [ "/path", "/path/to", "/path/to/folder", "/path/to/folder/folder" ],
+func directoryDiff(prev, current string) []string {
```
[the remark this answers is not in the document]
Comment (CollinShoop) [reply registry/storage/driver/s3-aws/s3.go:1195]: Will take a look at that 👍🏻

File: registry/storage/driver/s3-aws/s3.go (hunk 4 above)
Comment (waynr) [reply registry/storage/driver/s3-aws/s3.go:1132]: Ah, I was misunderstanding the break logic in `directoryDiff` - I see how it works now.

File: registry/storage/driver/s3-aws/s3.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (CollinShoop) [reply registry/storage/driver/s3-aws/s3.go:1195]: I'm not sure how `filepath.Rel` and `SplitList` would help here, though I can replace `sort.Sort` with something to reverse the list ordering, as the way this is done generates a list in reverse order compared to how we want them to be walked. If you have an idea of an alternate implementation, do you mind writing it up?

[the remark this answers is not in the document]
Comment (CollinShoop) [reply registry/storage/driver/s3-aws/s3.go:1195]: Using a simple `reverse` function now and all the unit tests still pass 👍🏻

File: registry/storage/driver/s3-aws/s3.go [hunk 6]
```diff
@@ -1082,110 +1083,87 @@ func (d *driver) Walk(ctx context.Context, from string, f storagedriver.WalkFn)
...
+	// for extreme edge cases but for the general use case in a registry, this is orders of magnitude
+	// faster than a more explicit recursive implementation.
+	listObjectErr := s.ListObjectsV2PagesWithContext(ctx, listObjectsInput, func(objects *s3.ListObjectsV2Output, lastPage bool) bool {
+		walkInfos := make([]storagedriver.FileInfoInternal, 0, len(objects.Contents))
 
 		for _, file := range objects.Contents {
-			walkInfos = append(walkInfos, walkInfoContainer{
+			filePath := strings.Replace(*file.Key, d.s3Path(""), prefix, 1)
+
+			// get a list of all inferred directories skipped between the previous directory and this file
```
Comment (adamwg) [inline registry/storage/driver/s3-aws/s3.go:1122]: Nit: "skipped" is confusing, since we're not implementing skipping in the `ErrSkipDir` sense here. I think the comment would be clear if you just omit the word "skipped" - we're finding all the inferred directories.

File: registry/storage/driver/s3-aws/s3.go [hunk 7]
```diff
@@ -1082,110 +1083,87 @@ func (d *driver) Walk(ctx context.Context, from string, f storagedriver.WalkFn)
...
-		}
+	// The s3 list object API will list all objects under the starting path recursively, in sorted,
+	// depth-first ordering https://docs.aws.amazon.com/AmazonS3/latest/userguide/ListingKeysUsingAPIs.html
+	// S3 does not return directory without the "delimiter":"/" argument passed, but they can to be
+	// inferred from object paths and then de-duplicated to avoid calling the walk fn redundantly.
+	// With files returned in sorted depth-first order, directories are inferred in the same order,
+	// so avoiding redundant directories is a matter of not passing the same directory to the walk fn twice in a row.
+	// ErrSkipDir is handled by explicitly skipping over any files under the skipped directory. This may sub-optimal
+	// for extreme edge cases but for the general use case in a registry, this is orders of magnitude
+	// faster than a more explicit recursive implementation.
```
Comment (adamwg) [inline registry/storage/driver/s3-aws/s3.go:1115]: Nit: It wasn't initially clear to me from this comment that the list is recursive *only* when `Delimiter` is not passed. I would reword a bit:

```suggestion
	// When the "delimiter" argument is omitted, the S3 list API will list all objects in the bucket
	// recursively, omitting directory paths. Objects are listed in sorted, depth-first order so we
	// can infer all the directories by comparing each object path to the last one we saw.
	// See: https://docs.aws.amazon.com/AmazonS3/latest/userguide/ListingKeysUsingAPIs.html
	//
	// With files returned in sorted depth-first order, directories are inferred in the same order,
	// so avoiding redundant directories is a matter of not passing the same directory to the walk fn twice in a row.
	// ErrSkipDir is handled by explicitly skipping over any files under the skipped directory. This may be sub-optimal
	// for extreme edge cases but for the general use case in a registry, this is orders of magnitude
	// faster than a more explicit recursive implementation.
```

**Review by adamwg — APPROVED** [review]
This looks great, though it took me a couple of read-throughs (and some S3 doc reading) to understand the correctness. I've left a couple of comment suggestions that would help make it more obvious for the reader.

File: registry/storage/driver/s3-aws/s3.go (hunk 7 above)
Comment (CollinShoop) [reply registry/storage/driver/s3-aws/s3.go]: Thanks 👍🏻

File: registry/storage/driver/s3-aws/s3.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (waynr) [reply registry/storage/driver/s3-aws/s3.go:1195]: > I'm not sure how filepath.Rel and SplitList would help here

The main benefit in my view would be improved readability and remove the need for the sort (or reverse now i guess).
